arangod/IResearch/IResearchAqlAnalyzer.cpp

```diff
@@ -0,0 +1,562 @@
+                errorMessage = "Invalid bind parameter found '";
+                errorMessage.append(parameterName).append("'");
+                return false;
+              }
+            } break;
+            // by default all is forbidden
+            default:
+              errorMessage = "Node type '";
+              errorMessage.append(node->getTypeString()) + "' is forbidden for aql analyzer";
+              return false;
+              break;
+          }
+          return true;
+        });
+    if (!errorMessage.empty()) {
+      return {TRI_ERROR_BAD_PARAMETER, errorMessage};
+    }
+  } catch (arangodb::basics::Exception const& e) {
+    return {TRI_ERROR_QUERY_PARSE, e.message()};
+  } catch (std::exception const& e) {
+    return {TRI_ERROR_QUERY_PARSE, e.what()};
+  } catch (...) {
+    TRI_ASSERT(FALSE);
+    return {TRI_ERROR_QUERY_PARSE, "Unexpected"};
+  }
+  return {};
+}
+
+irs::analysis::analyzer::ptr make_slice(VPackSlice const& slice) {
+  arangodb::iresearch::AqlAnalyzer::Options options;
+  if (parse_options_slice(slice, options)) {
+    auto validationRes =
+        validateQuery(options.queryString,
+                      arangodb::DatabaseFeature::getCalculationVocbase());
+    if (validationRes.ok()) {
+      return std::make_shared<arangodb::iresearch::AqlAnalyzer>(options);
+    } else {
+      LOG_TOPIC("f775e", WARN, arangodb::iresearch::TOPIC)
+          << "error validating calculation query: " << validationRes.errorMessage();
+    }
+  }
+  return nullptr;
+}
+}  // namespace
+
+namespace arangodb {
+namespace iresearch {
+
+/*static*/ bool AqlAnalyzer::normalize_vpack(const irs::string_ref& args, std::string& out) {
+  auto const slice = arangodb::iresearch::slice(args);
+  VPackBuilder builder;
+  if (normalize_slice(slice, builder)) {
+    out.resize(builder.slice().byteSize());
+    std::memcpy(&out[0], builder.slice().begin(), out.size());
+    return true;
+  }
+  return false;
+}
+
+/*static*/ bool AqlAnalyzer::normalize_json(const irs::string_ref& args,
+                                                    std::string& out) {
+  auto src = VPackParser::fromJson(args);
+  VPackBuilder builder;
+  if (normalize_slice(src->slice(), builder)) {
+    out = builder.toString();
+    return true;
+  }
+  return false;
+}
+
+/*static*/ irs::analysis::analyzer::ptr AqlAnalyzer::make_vpack(irs::string_ref const& args) {
+  auto const slice = arangodb::iresearch::slice(args);
+  return make_slice(slice);
+}
+
+
+/*static*/ irs::analysis::analyzer::ptr AqlAnalyzer::make_json(irs::string_ref const& args) {
+  auto builder = VPackParser::fromJson(args);
+  return make_slice(builder->slice());
+}
+
+AqlAnalyzer::AqlAnalyzer(Options const& options)
+  : irs::analysis::analyzer(irs::type<AqlAnalyzer>::get()),
+    _options(options),
+    _query(new CalculationQueryContext(arangodb::DatabaseFeature::getCalculationVocbase())),
+    _itemBlockManager(&_resourceMonitor, SerializationFormat::SHADOWROWS), 
```

Comment:
I think the default memory limit of a resourceMonitor is unbound.
Now every AQL analyzer will get it's own resourceMonitor and can potentially use unbounded amount of main memory.

It is unclear to me what happens if the user manages to write an Analyzer that triggers out of memory situation.
I guess this will potentially crash the server.
If the server reboots, will it find the same Analyzer again and retry => end up in out of memory again ?

Reply:
I do not see a way to limit maximum amount of used memory from outside